packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3233 @@
+    }
+
+    return visitChildren(this);
+  }
+
+  @override
+  int compareTo(_MenuNode other) {
+    final List<_MenuNode> allNodes = topLevel.parent!.descendants.toList();
+    return allNodes.indexOf(this).compareTo(allNodes.indexOf(other));
+  }
+
+  /// Returns the list of node ancestors with any of the ancestors that appear
+  /// in the [other]'s ancestors removed. Includes this node in the results.
+  List<_MenuNode> ancestorDifference(_MenuNode? other) {
+    final List<_MenuNode> myAncestors = <_MenuNode>[...ancestors, this];
+    final List<_MenuNode> otherAncestors = <_MenuNode>[
+      ...other?.ancestors ?? <_MenuNode>[],
+      if (other != null) other,
+    ];
+    int skip = 0;
+    for (; skip < myAncestors.length && skip < otherAncestors.length; skip += 1) {
+      if (myAncestors[skip] != otherAncestors[skip]) {
+        break;
+      }
+    }
+    return myAncestors.sublist(skip);
+  }
+
+  /// Get all of the registered children of the given menu that are focusable.
+  /// Used for menu traversal.
+  List<_MenuNode> get focusableChildren {
+    return children.where((_MenuNode child) => child.focusNode?.canRequestFocus ?? false).toList();
+  }
+
+  /// Called whenever this menu is opened by being set as the
+  /// [_MenuBarController.openMenu].
+  ///
+  /// Used to avoid calling [MenuItem.onOpen] unnecessarily.
+  void open() {
+    if (isOpen) {
+      return;
+    }
+    isOpen = true;
+    item.onOpen?.call();
+  }
+
+  /// Called whenever this menu is closed by another menu being set as the
+  /// [_MenuBarController.openMenu].
+  ///
+  /// Used to avoid calling [MenuItem.onClose] unnecessarily.
+  void close() {
+    if (!isOpen) {
+      return;
+    }
+    isOpen = false;
+    item.onClose?.call();
+  }
+
+  // Used for testing to verify which item this is.
+  @override
+  String toStringShort({DiagnosticLevel minLevel = DiagnosticLevel.info}) {
+    return item.toStringShort();
+  }
+
+  @override
+  List<DiagnosticsNode> debugDescribeChildren() {
+    return <DiagnosticsNode>[
+      ...children.map<DiagnosticsNode>((_MenuNode item) => item.toDiagnosticsNode()),
+    ];
+  }
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<MenuItem>('item', item));
+    properties.add(DiagnosticsProperty<_MenuNode>('parent', parent, defaultValue: null));
+    properties.add(IntProperty('numChildren', children.length, defaultValue: null));
+  }
+}
+
+/// A controller that allows control of a [MenuBar].
+///
+/// Normally, it's not necessary to create a `MenuBarController` to use a
+/// [MenuBar], but if you need to be able to close any open menus with the
+/// [closeAll] method in response to an event, you can create one and pass it to
+/// the [MenuBar].
+///
+/// If the place you wish to close all the menus (for instance, when a control
```

Comment:
the first half of this sentence seems to be missing something (or at least I don't understand what it is trying to say)